private/exponential_backoff.hpp

```diff
@@ -11,51 +11,94 @@
+   *
+   * @param mul_coeff Multiplier coefficient used in the exponential delay calculation.
    * @param max_delay The upper bound for all the backoff delays.
    */
-  ExponentialBackoff(std::chrono::milliseconds initial_delay, std::chrono::milliseconds max_delay)
-      : initial_delay_(initial_delay), max_delay_(max_delay) {}
+  ExponentialBackoff(std::chrono::milliseconds mul_coeff, std::chrono::milliseconds max_delay)
+      : mul_coeff_(mul_coeff), max_delay_(max_delay) {
+    if ((mul_coeff <= std::chrono::milliseconds::zero()) ||
+        (max_delay <= std::chrono::milliseconds::zero())) {
+      throw AstarteInvalidInputException(
+          "Negative or zero duration passed to the ExponentialBackoff constructor");
+    }
+    if (max_delay < mul_coeff) {
+      throw AstarteInvalidInputException(
+          "Maximum delay should be larger or equal to the initial delay");
+    }
+  }
 
   /**
    * @brief Calculate and returns the next backoff delay.
-   * @details Computes the appropriate delay for the current backoff generation and increments the
-   * internal generation counter for the next call.
+   * @details See the documentation of the constructor of this class for an explanation on how
+   * this delay is computed.
    * @return The calculated delay duration.
    */
   auto getNextDelay() -> std::chrono::milliseconds {
-    constexpr double BACKOFF_FACTOR = 2.0;
+    const ChronoMillisRep mul_coeff = mul_coeff_.count();
+    const ChronoMillisRep max_milliseconds = std::chrono::milliseconds::max().count();
+    const ChronoMillisRep max_allowed_final_delay = max_milliseconds - mul_coeff;
 
-    const auto initial_delay_ms = static_cast<double>(initial_delay_.count());
+    // Update last delay value with the new value
+    ChronoMillisRep delay = 0;
+    if (prev_delay_ == 0) {
+      delay = mul_coeff;
+    } else if (prev_delay_ <= (max_allowed_final_delay / 2)) {
+      delay = 2 * prev_delay_;
+    } else {
+      delay = max_allowed_final_delay;
+    }
 
-    const auto delay_ms = initial_delay_ms * std::pow(BACKOFF_FACTOR, generated_delays_);
-    // Apply a positive jitter (a random value between 0 and initial_delay_)
-    const auto jitter = dist_(gen_) * initial_delay_ms;
+    // Bound the delay to the maximum
+    ChronoMillisRep bounded_delay = std::min(delay, max_delay_.count());
 
-    const auto total_delay_ms = static_cast<int64_t>(delay_ms + jitter);
-    const auto jittery_delay = std::chrono::milliseconds(total_delay_ms);
+    // Store the new delay before jitter application
+    prev_delay_ = bounded_delay;
 
-    generated_delays_++;
+    // Insert some jitter
+    ChronoMillisRep jitter_minimum = -mul_coeff;
+    if (bounded_delay - mul_coeff < 0) {
+      jitter_minimum = 0;
+    }
+    ChronoMillisRep jitter_maximum = mul_coeff;
+    if (bounded_delay > max_milliseconds - mul_coeff) {
+      jitter_maximum = max_milliseconds - bounded_delay;
```

Comment:
Is it possible for bounded_delay and max_milliseconds to be equal? 
Since I can set `max_delay = std::chrono::milliseconds::max()`.
In that case if these are equal, the difference is zero and the jitter_maximum is zero.

Reply:
Yes, that should still work fine. As the positive jitter will be 0, but the negative will still be `-mul_coeff`.
So, some jitter would always be applied.